core/derive/src/lib.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+extern crate proc_macro;
+
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
+
+synstructure::decl_derive!([Flush] => flush_derive);
+synstructure::decl_derive!([AllocateUsing] => allocate_using_derive);
+
+pub(crate) fn flush_derive(mut s: synstructure::Structure) -> TokenStream2 {
+    s.bind_with(|_| synstructure::BindStyle::Move);
+    s.add_bounds(synstructure::AddBounds::Fields);
+    let mut requires_match = false;
+    let body = s.each(|bi| {
+        requires_match = true;
+        quote! {
+            ink_core::storage::Flush::flush(#bi)
+        }
+    });
+    let body = if requires_match {
+        quote! {
+            match self {
+                #body
+            }
+        }
+    } else {
+        quote! {}
+    };
+    s.gen_impl(quote! {
+        extern crate ink_core;
+
+        gen impl ink_core::storage::Flush for @Self {
+            fn flush(&mut self) {
+                #body
+            }
+        }
+    })
+}
+
+pub(crate) fn allocate_using_derive(mut s: synstructure::Structure) -> TokenStream2 {
+    // We cannot implement AllocateUsing on enums because we cannot specify
+    // which variant we are going to use.
+    if let syn::Data::Enum(ref _enum_data) = s.ast().data {
+        panic!("cannot derive AllocateUsing for enums")
+    }
+    s.bind_with(|_| synstructure::BindStyle::Move);
+    s.add_bounds(synstructure::AddBounds::Fields);
+    let body = s.variants()[0].construct(|field, _| {
```

Comment:
Would be good to have a comment here that structs have a single "variant". Or even assert that `s.variants().len() == 0` - just to make it clear what's going on here.

Reply:
Single variant is already guaranteed by the design of `synstructure` and the fact that I added a check to disallow `enum` (that is the only way to have multiple variants). The design of `synstructure` treats everything as `enum` which is brilliant but confusing at first sight. I will add some docs that explains this.